utilities/include/vec4.hpp

```diff
@@ -0,0 +1,171 @@
+#ifndef VEC4_HPP
+#define VEC4_HPP
+
+#include <cstddef>
+#include <ostream>
+#include <cmath>
+#include <xsimd/xsimd.hpp>
+#include <type_traits>
+
+#include "config.hpp"
+
+/*  *Macros provided by config.hpp
+    - ALIGN_WIDTH: Alignment width for class, since it is dependent on Arch.
+    - UArch: Architecture to be used.
+    - UType: Data type to be used.
+    - CONSTEXPR_CMATH: Switch constexpr if available in cmath.
+*/
+class alignas(ALIGN_WIDTH) Vec4{
+    
+    public:
+        constexpr Vec4() : m_v{0,0,0,0} {}
+        constexpr explicit Vec4(Utype x) : m_v{x,x,x,x} {}
+        constexpr Vec4(Utype x,Utype y,Utype z,Utype w) : m_v{x,y,z,w} {}
+        constexpr explicit Vec4(xsimd::batch<Utype,UArch> x): m_v{0,0,0,0} {x.store_aligned(m_v);}
+        
+        //getters
+        constexpr Utype x() const{
+            return m_v[0];
+        }
+
+        constexpr Utype y() const{
+            return m_v[1];
+        }
+        
+        constexpr Utype z() const{
+            return m_v[2];
+        }
+
+        constexpr Utype w() const{
+            return m_v[3];
+        }
+        
+        //indexing operations
+        constexpr const Utype& operator[](std::size_t i) const{
+            return m_v[i];
+        }
+
+        constexpr Utype& operator[](std::size_t i){
+            return m_v[i];
+        }
+
+        // Common Vec operations
+        constexpr Utype dot(const Vec4& v2) const{
+            if (std::is_constant_evaluated()){
+                return (*this).x()*v2.x() + (*this).y()*v2.y() + (*this).z()*v2.z() + (*this).w()*v2.w() ;
+            } else {    
+                xsimd::batch<Utype,UArch> temp = vec2batch(*this)*vec2batch(v2);
+                return sum(temp);
+            }
+        }
+        
+        CONSTEXPR_CMATH Utype norm() const{
+            return sqrt(dot(*this));
+        }
+
+        CONSTEXPR_CMATH static Utype angle(const Vec4& v1,const Vec4& v2){
+            return acos(v1.dot(v2)/(v1.norm()*v2.norm()));
+        }
+        CONSTEXPR_CMATH Vec4 unit() const{
+            return *this/this->norm();
+        }
+
+        // Negation 
+        friend constexpr Vec4 operator-(const Vec4& v1){
+            if (std::is_constant_evaluated()){
+                return Vec4(-v1.x(),-v1.y(),-v1.z(),-v1.w());
+            } else {
+                return Vec4(-vec2batch(v1));
+            }
+        }
+
+        // Addition and Substraction
+        friend constexpr Vec4& operator+=(Vec4& v1, const Vec4& v2){
+            if (std::is_constant_evaluated()){
+                v1.m_v[0] += v2.x();
+				v1.m_v[1] += v2.y();
```

Comment:
Funky formatting; fix it please.

Reply:
Also why are these `friend`s? Once again, consistency is most important here. Either every function should be a friend, or none of them should be. (In particular, `cross` and `dot`.)

Reply:
This looks alright to me in vs code. Since they are declared inside the class it would be better if I just unfriend them.